### PR TITLE
Issue3579 - Check for CTRL-C During sleep on esp32s2

### PR DESCRIPTION
**Pull request by DavePutz**
This PR is not very elegant. But, in its defense, the following fixes were tried without success:
1) I tried to find a way to interrupt or terminate vTaskDelayUntil(). However, nothing I found was able to do that. Trying to
    do anything about the sleep state during another interrupt handler (such as ticks or usb) generally resulted in either a 
   complete hang or did nothing.
2) The esp32s2 light sleep state was investigated. However, any light sleep past about 1 second resulted in the USB session 
  getting disconnected, which makes getting a CTRL-C in impossible. I did not look at the effect of light sleep on Wifi or BLE, but 
  there is a comment in the code indicating those might be a problem as well.  I did not discover any way to keep the two pins 
  needed for USB alive during light sleep.
3) The sleep interval in the PR (1) was chosen to maintain as accurate a sleep time as possible. Higher values grew increasingly 
   inaccurate for a time.sleep() call.
Any comments and suggestions for improvement are quite welcome.

### CODE REVIEW

**Comment by tannewt** [comment]
I think you could invert the API you need from FreeRTOS. Instead of using a time delay API, use one that is blocked on inter-task flag with a time out. xEventGroupWaitBits is what you'd need I think. https://www.freertos.org/xEventGroupWaitBits.html Then, share a group of event bits with TinyUSB callbacks that can indicate a ctrl-c. Then wait on the event bit while timing out with the sleep time.

**Comment by DavePutz** [comment]
That sounds like a better fix than a clumsy loop. I will investigate and
see what I can do with the suggestion.

**Comment by DavePutz** [comment]
Reworked the PR to use RTOS task notification instead of a loop.

File: ports/esp32s2/supervisor/usb.c [hunk 1]
```diff
@@ -52,6 +52,8 @@
 StackType_t  usb_device_stack[USBD_STACK_SIZE];
 StaticTask_t usb_device_taskdef;
 
+TaskHandle_t xTaskToNotify = NULL;
```
Comment (tannewt) [inline ports/esp32s2/supervisor/usb.c:55]:
```suggestion
TaskHandle_t sleeping_circuitpython_task = NULL;
```

File: ports/esp32s2/supervisor/port.c [hunk 2]
```diff
@@ -188,25 +190,28 @@ void port_disable_tick(void) {
...
-    // esp_light_sleep_start()
+
     if (sleep_time_duration == 0) {
         return;
     }
-    vTaskDelayUntil(&sleep_time_set, sleep_time_duration);
+    notify_wait = xTaskNotifyWait(0x01,0x01,&NotifyValue,
+                             sleep_time_duration );
+    if (NotifyValue == 1) {
+      xTaskToNotify = NULL;
```
Comment (tannewt) [inline ports/esp32s2/supervisor/port.c:210]:
```suggestion
      sleeping_circuitpython_task = NULL;
```

File: ports/esp32s2/supervisor/port.c [hunk 3]
```diff
@@ -188,25 +190,28 @@ void port_disable_tick(void) {
     esp_timer_stop(_tick_timer);
 }
 
-TickType_t sleep_time_set;
 TickType_t sleep_time_duration;
+uint32_t NotifyValue = 0;
+BaseType_t notify_wait = 0;
```
Comment (tannewt) [inline ports/esp32s2/supervisor/port.c:195]: Why do these need to be outside the function?

File: ports/esp32s2/supervisor/port.c [hunk 4]
```diff
@@ -54,6 +55,7 @@
 
 uint32_t* heap;
 uint32_t heap_size;
+extern TaskHandle_t xTaskToNotify;
```
Comment (tannewt) [inline ports/esp32s2/supervisor/port.c:58]: Could this go in `ports/esp32s2/supervisor/port.h`?

```suggestion
extern TaskHandle_t sleeping_circuitpython_task;
```

File: ports/esp32s2/supervisor/port.c [hunk 5]
```diff
@@ -188,25 +190,28 @@ void port_disable_tick(void) {
...
 TickType_t sleep_time_duration;
+uint32_t NotifyValue = 0;
+BaseType_t notify_wait = 0;
+
 void port_interrupt_after_ticks(uint32_t ticks) {
-    sleep_time_set = xTaskGetTickCount();
-    sleep_time_duration = ticks / portTICK_PERIOD_MS;
-    // esp_sleep_enable_timer_wakeup(uint64_t time_in_us)
+    sleep_time_duration = (ticks * 100)/1024;
+    xTaskToNotify = xTaskGetCurrentTaskHandle();
```
Comment (tannewt) [inline ports/esp32s2/supervisor/port.c:199]:
```suggestion
    sleeping_circuitpython_task = xTaskGetCurrentTaskHandle();
```

File: ports/esp32s2/supervisor/usb.c [hunk 6]
```diff
@@ -114,3 +116,21 @@ void init_usb_hardware(void) {
...
+ */
+void tud_cdc_rx_wanted_cb(uint8_t itf, char wanted_char)
+{
+    (void) itf; // not used
+    // Workaround for using lib/utils/interrupt_char.c
+    // Compare mp_interrupt_char with wanted_char and ignore if not matched
+    if (mp_interrupt_char == wanted_char) {
+        tud_cdc_read_flush();    // flush read fifo
+        mp_keyboard_interrupt();
+        if (xTaskToNotify != NULL) {
```
Comment (tannewt) [inline ports/esp32s2/supervisor/usb.c:132]:
```suggestion
        // CircuitPython's VM is run in a separate FreeRTOS task from TinyUSB. So, we must notify the
        // other task when a CTRL-C is received.
        if (sleeping_circuitpython_task != NULL) {
```

File: ports/esp32s2/supervisor/usb.c [hunk 7]
```diff
@@ -114,3 +116,21 @@ void init_usb_hardware(void) {
...
+          xTaskNotifyGive(xTaskToNotify);
```
Comment (tannewt) [inline ports/esp32s2/supervisor/usb.c:133]:
```suggestion
          xTaskNotifyGive(sleeping_circuitpython_task);
```

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks @DavePutz! A couple questions and a variable rename suggestion to better separate FreeRTOS APIs from our own.

I suspect we'll need to use this for auto-reload during sleep as well.

**Comment by DavePutz** [comment]
Suggested changes implemented except for moving the extern declare to ports/esp32s2/supervisor/port.h. That file does not exist; is it worth creating it to hold this one line?

**Comment by tannewt** [comment]
> Suggested changes implemented except for moving the extern declare to ports/esp32s2/supervisor/port.h. That file does not exist; is it worth creating it to hold this one line?

Up to you. I suspect we'll need access to it to fix the auto-reload while sleeping too but that fix might live in these files too.

**Review by tannewt — APPROVED** [review]
Looks good! Thank you!

**Comment by DavePutz** [comment]
Since you think it may be useful in the future I have split out the extern declare. I did name it ports/esp32s2/supervisor/esp_port.h, however, to avoid naming conflicts with the global supervisor/port.h

**Review by tannewt — APPROVED** [review]
Thanks!

**Pull request merged**